File: KotlinPractice/src/main/java/com/android/kotlinpractice/types/classes/java/Person.java
package com.android.kotlinpractice.types.classes.java;

import com.android.kotlinpractice.types.classes.kotlin.SimpleInterface;

/**
 * TODO
 *
 * @author dev.liang <a href="mailto:dev.liang@outlook.com">Contact me.</a>
 * @version 1.0
 * @since 2020/01/29 16:34
 */
public class Person implements SimpleInterface {

    private String name;
    private int age;
    private int sex;


    public Person() {
    }

    public Person(String name, int age, int sex) {
        this.name = name;
        this.age = age;
        this.sex = sex;
    }


    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public int getSex() {
        return sex;
    }

    public void setSex(int sex) {
        this.sex = sex;
    }

    @Override
    public void simpleMethod() {

    }

    @Override
    public int getSimpleProperty() {
        return 0;
    }

    @Override
    public void setSimpleProperty(int i) {

    }
}

File: KotlinPractice/src/main/java/com/android/kotlinpractice/types/classes/kotlin/Person.kt
package com.android.kotlinpractice.types.classes.kotlin


/**
 * TODO
 *
 * 这里定义的 var age、name、sex 和 java 中定义的变量不同，java 是 field，kotlin 中是 Property
 * Property 相当于 java 中的 field + get + set
 *
 * 这里的 get 和 set 默认可以不写，也可以自定义书写
 *
 * @author dev.liang <a href="mailto:dev.liang@outlook.com">Contact me.</a>
 * @version 1.0
 * @since 2020/01/28 11:48
 */
class Person(age: Int, name: String, sex: Int) :AbstractClass(),SimpleInterface {

    override fun absMethod() {
    }

    override fun simpleMethod() {

    }

    override var simpleProperty: Int
//        get() = 3
        get() {
            return 3
        }
        set(value) {}


    var age: Int = age
        get() {
            return field // field 相当于 backing field
        }
        set(value) {
            field = value
        }

    var name: String = name
    var sex: Int = sex


}


fun main() {
    var person = Person(18,"dev",0)
    // kotlin 的 属性引用；和 函数引用很类似
    val ageReference = Person::age // 未绑定 Receiver
    ageReference.set(person, 29)
    println("ageReference = ${ageReference.get(person)}")

    val nameReference = person::name  // 绑定 Receiver （person）
    nameReference.set("dev.liang")
    println("nameReference = ${nameReference.get()}")



}







